pkg/kubelet/eviction/eviction_manager.go

```diff
@@ -405,6 +410,11 @@ func (m *managerImpl) synchronize(diskInfoProvider DiskInfoProvider, podFunc Act
 	// we kill at most a single pod during each eviction interval
 	for i := range activePods {
 		pod := activePods[i]
+		if m.podsEvicted.Has(string(pod.UID)) {
+			klog.InfoS("Eviction manager: not evicting pod that we already attempted to evict: ", "namespace", pod.Namespace, "pod", pod.Name)
+			continue
```

Comment:
Does this mean we're trying to evict certain pod only once and not going to repeat evicting it? Should we instead change ranking so pods that are not yet evicted would come first?

Reply:
We could do that, but we still would want to check that the pod we want to evict hasn't already been evicted (eg. if there is only one pod on the node, we don't want to re-evict it)

Reply:
As the current logic is different (we re-evicting pods, right?), I agree with @kannon92  that this needs to be discussed on the SIG-Node meeting. Additionally, it would be good to explicitly state this in the PR description.

Reply:
I came to the meeting, but there wasn't enough time for my item, I'll try the next one. Right now, we aren't re-evicting because we just wait for the one pod that we're evicting to finish shutting down before running through the loop again

Reply:
I added this to the PR description

Reply:
> Right now, we aren't re-evicting because we just wait for the one pod that we're evicting to finish shutting down before running through the loop again

Can you point me out where this waiting is happening?
If this is the case, how adding pod to the `podEvicted` would help to prevent that waiting?

Reply:
So [we go from here](https://github.com/kubernetes/kubernetes/blob/016d8b11ae250ca933fe83de312bbe7c4945cd20/pkg/kubelet/eviction/eviction_manager.go#L603) to [here](https://github.com/kubernetes/kubernetes/blob/d9c54f69d4bb7ae1bb655e1a2a50297d615025b5/pkg/kubelet/pod_workers.go#L1666-L1677) where we wait up till the graceful termination period 

My addition of [killPodNowWithMaxTimeout](https://github.com/kubernetes/kubernetes/pull/124081/files#diff-c6d21f5481c391a3f32bcfddf45c9c20132b2876e8be9690289474b76b71d809) and its [usage here](https://github.com/kubernetes/kubernetes/pull/124081/files#diff-67dd9e8f3ee257072765326cb4f242852554a2c0753563fa51e292c0a63a7b94) means we will wait for only 5m for the pod to shut down before moving on

Reply:
Thanks, now it's much more clear to me.

Reply:
Yea, killPod is blocking. My main questions/designs are do we want to override long graceperiods for eviction?

If one has pods that take a long time to kill, then why wouldn't this end up killing most of the active pods? If every pod takes hours to kill, then I see the eviction manager looping over all the active pods and killing them.

Reply:
My thinking is that the priority is to protect the node, so rather than letting it run out of resources, evict more pods. And if resource usage keeps climbing in the meantime, it will switch to hard evicts (these don't respect grace period). Alternatively, we can just apply this logic to when we reach the hard evict threshold. (Eg. only soft evict one at a time, but once we get past the hard threshold, start hard evictions even if the soft eviction is not done)

Reply:
So, if someone wants to kill all pods on the node, it would be enough to create a new one with huge soft grace timeout consuming a lot of resources?

Reply:
So in your cluster, are you seeing that disk usage is hovering between soft and hard?

Reading the design doc on eviction, I believe that we should do evictionHard if those limits are hit?

Reply:
> So, if someone wants to kill all pods on the node, it would be enough to create a new one with huge soft grace timeout consuming a lot of resources?

Yeah, I suppose so. Without this change, doing that would probably still kill all pods once resources are exhausted

>So in your cluster, are you seeing that disk usage is hovering between soft and hard?

No, this is something I tested for, but did not actually experience first hand

> Reading the design doc on eviction, I believe that we should do evictionHard if those limits are hit?

I assumed that as well! But nope, it just continues to wait on the soft evicted pod to terminate

Reply:
Note that hard evictions don't work at all right now, pods still get their full termination grace period. The fix is https://github.com/kubernetes/kubernetes/pull/124063

Reply:
> Yea, killPod is blocking.

@kannon92 Can you explain how hard eviction is done when eviction manager is waiting for killPodFunc blocked by huge soft grace period? I looked at the code and it seems to be synchronous.

Reply:
> Note that hard evictions don't work at all right now, pods still get their full termination grace period.

That was my guess too :(

> The fix is https://github.com/kubernetes/kubernetes/pull/124063

That looks like unconditional overriding of grace period to me so far. Not sure I understand how that would work. I'll look again.